kimchi/src/circuits/polynomials/chacha.rs

```diff
@@ -180,270 +182,282 @@ pub fn xor_table<F: Field>() -> Vec<Vec<F>> {
     res
 }
 
-// This is just for tests. It doesn't set up the permutations
-pub fn chacha20_gates() -> Vec<GateType> {
-    let mut gs = vec![];
-    for _ in 0..20 {
-        use GateType::*;
-        for _ in 0..4 {
-            for &g in &[ChaCha0, ChaCha1, ChaCha2, ChaCha0, ChaChaFinal] {
-                gs.push(g);
-                gs.push(Zero);
+//
+// Implementation internals
+//
+
+/// 8-nybble sequences that are laid out as 4 nybbles per row over the two row,
+/// like y^x' or x+z
+fn chunks_over_2_rows<F>(col_offset: usize) -> Vec<E<F>> {
+    (0..8)
+        .map(|i| {
+            let r = if i < 4 { Curr } else { Next };
+            witness(col_offset + (i % 4), r)
+        })
+        .collect()
+}
+
+/// Creates a constraint to enforce that b is either 0 or 1.
+fn boolean<F: Field>(b: &E<F>) -> E<F> {
```

Comment:
I would be tempted to move this to `expr` as well, since it's only related to chacha in the sense that its gate uses it, but we use it in several other gates.

Reply:
ok, implementing this in another PR ;)

Reply:
https://github.com/o1-labs/proof-systems/pull/290

Reply:
I don't like that that PR comes *after* this one, since it leads to 'approving' a change that I don't approve of. In future, I would prefer if the equivalent of #290 was against master and the equivalent of this PR was built on top, by merging the new PR's branch and fixing any changes.